frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -148,40 +156,65 @@ import {
 import { proximityDate, MINS_MILLIS } from '@model/contracts/shared/time.js'
 import { cloneDeep, debounce, throttle, delay } from 'turtledash'
 import { EVENT_HANDLED } from '@chelonia/lib/events'
 import { compressImage } from '@utils/image.js'
 import { swapMentionIDForDisplayname, makeMentionFromUserID } from '@model/chatroom/utils.js'
+import DynamicScroller from '@components/vue-virtual-scroller/DynamicScroller.vue'
+import DynamicScrollerItem from '@components/vue-virtual-scroller/DynamicScrollerItem.vue'
 
 const ignorableScrollDistanceInPixel = 500
 
+const enqueue = function (fn) {
+  const chatroomID = this.ephemeral.renderingChatRoomId
+  return sbp('okTurtles.eventQueue/queueEvent', CHATROOM_EVENTS, () => {
+    if (this.chatroomHasSwitchedFrom(chatroomID)) return
+
+    return fn()
+  })
+}
+
+// const xdebounce = (x, ..._) => x
+
 // The following methods are wrapped inside `debounce`, which requires calling
 // flush before the references used go away, like when switching groups.
 // Vue.js binds methods, which means that properties like `.flush` become
 // inaccessible. So, instead we define these methods outside the component and
 // manually bind them in `mounted`.
-const onChatScroll = function () {
+const onChatScroll = function (ev) {
+  console.error('@@@@onChatScroll', ev, !this.$refs.conversation, !this.summary.isJoined, 'mI', this.ephemeral.messagesInitiated)
   // NOTE: Should be careful of using the currentChatRoomState
   //       since those states are depends on the currentChatRoomId, not renderingChatRoomId
-  if (!this.$refs.conversation || !this.summary.isJoined) {
+  if (!this.$refs.conversation || !this.summary.isJoined || !this.ephemeral.messages.length) {
     return
   }
 
-  const curScrollTop = this.$refs.conversation.scrollTop
-  const curScrollBottom = curScrollTop + this.$refs.conversation.clientHeight
-  const scrollTopMax = this.$refs.conversation.scrollHeight - this.$refs.conversation.clientHeight
+  const curScrollTop = this.$refs.conversation.$el.scrollTop
+  // const curScrollBottom = curScrollTop + this.$refs.conversation.$el.clientHeight
+  const scrollTopMax = this.$refs.conversation.$el.scrollHeight - this.$refs.conversation.$el.clientHeight
+  const scroller = this.$refs.conversation.$refs.scroller
   this.ephemeral.scrollableDistance = scrollTopMax - curScrollTop
+  // const scroller = this.$refs.conversation.$refs.scroller
+
+  if (curScrollTop < 5) {
+    this.onScrollStart()
+  } else if (curScrollTop + 5 > scrollTopMax) {
+    this.onScrollEnd()
+  }
 
-  for (let i = this.messages.length - 1; i >= 0; i--) {
-    const msg = this.messages[i]
-    if (msg.pending || msg.hasFailed) continue
-    const offsetTop = this.$refs[msg.hash][0].$el.offsetTop
+  // for (let i = this.ephemeral.messages.length - 1; i >= 0; i--) {
+  for (let i = (scroller?.$_endIndex ?? this.ephemeral.messages.length) - 1; i >= (scroller?.$_startIndex ?? 0); i--) {
```

Comment:
I could be wrong but `$_endIndex` and `$_startIndex` are the features  from `vue-virtual-scroller` package which apparently was later replaced by `DynamicScroller.vue` you wrote. If so, I think these should be removed.